drivers/at86rf215/at86rf215_netdev.c

```diff
@@ -0,0 +1,922 @@
+    }
+
+    return res;
+}
+
+static void _enable_tx2rx(at86rf215_t *dev)
+{
+    uint8_t amcs = at86rf215_reg_read(dev, dev->BBC->RG_AMCS);
+
+    /* disable AACK, enable TX2RX */
+    amcs |=  AMCS_TX2RX_MASK;
+    amcs &= ~AMCS_AACK_MASK;
+
+    at86rf215_reg_write(dev, dev->BBC->RG_AMCS, amcs);
+}
+
+static void _tx_end(at86rf215_t *dev, netdev_event_t event)
+{
+    netdev_t *netdev = (netdev_t *)dev;
+
+    /* listen to non-ACK packets again */
+    if (dev->flags & AT86RF215_OPT_ACK_REQUESTED) {
+        dev->flags &= ~AT86RF215_OPT_ACK_REQUESTED;
+        at86rf215_filter_ack(dev, false);
+    }
+
+    at86rf215_tx_done(dev);
+
+    if (dev->flags & AT86RF215_OPT_TELL_TX_END) {
+        netdev->event_callback(netdev, event);
+    }
+
+    dev->timeout = 0;
+    dev->state = AT86RF215_STATE_IDLE;
+}
+
+static void _ack_timeout_cb(void* arg) {
+    at86rf215_t *dev = arg;
+    dev->timeout = AT86RF215_TIMEOUT_ACK;
+    msg_send_int(&dev->timer_msg, dev->timer_msg.sender_pid);
+}
+
+static void _backoff_timeout_cb(void* arg) {
+    at86rf215_t *dev = arg;
+    dev->timeout = AT86RF215_TIMEOUT_CSMA;
+    msg_send_int(&dev->timer_msg, dev->timer_msg.sender_pid);
+}
+
+static void _set_idle(at86rf215_t *dev)
+{
+    dev->state = AT86RF215_STATE_IDLE;
+
+    uint8_t next_state;
+
+    if (dev->flags & AT86RF215_OPT_TX_PENDING) {
+        next_state = CMD_RF_TXPREP;
+    } else {
+        next_state = CMD_RF_RX;
+    }
+
+    at86rf215_rf_cmd(dev, next_state);
+}
+
+/* wake up the radio thread after ACK timeout */
+static void _start_ack_timer(at86rf215_t *dev)
+{
+    dev->timer_msg.type = NETDEV_MSG_TYPE_EVENT;
+    dev->timer_msg.sender_pid = thread_getpid();
+
+    dev->timer.arg = dev;
+    dev->timer.callback = _ack_timeout_cb;
+
+    xtimer_set(&dev->timer, dev->ack_timeout_usec);
+}
+
+/* wake up the radio thread after CSMA backoff period */
+static void _start_backoff_timer(at86rf215_t *dev)
+{
+    uint8_t be;
+    uint32_t base;
+
+    at86rf215_get_random(dev, &base, sizeof(base));
+
+    be = ((dev->csma_retries_max - dev->csma_retries) - 1) + dev->csma_minbe;
+
+    if (be > dev->csma_maxbe) {
+        be = dev->csma_maxbe;
+    }
+
+    uint32_t csma_backoff_usec = ((1LU << be) - 1) * dev->csma_backoff_period;
+    /* limit the 32bit random value to the current backoff */
+    csma_backoff_usec = base % csma_backoff_usec;
+
+    DEBUG("Set CSMA backoff to %lu (be %u min %u max %u base: %lu)\n",
+          csma_backoff_usec, be, dev->csma_minbe, dev->csma_maxbe, base);
+
+    dev->timer_msg.type = NETDEV_MSG_TYPE_EVENT;
+    dev->timer_msg.sender_pid = thread_getpid();
+
+    dev->timer.arg = dev;
+    dev->timer.callback = _backoff_timeout_cb;
+
+    xtimer_set(&dev->timer, csma_backoff_usec);
+}
+
+static inline bool _ack_frame_received(at86rf215_t *dev)
+{
+    /* check if the sequence numbers match */
+    return at86rf215_reg_read(dev, dev->BBC->RG_FBRXS + 2)
+        == at86rf215_reg_read(dev, dev->BBC->RG_FBTXS + 2);
```

Comment:
magic number?

Reply:
offset of the sequence number in an 802.15.4 ACK frame - I've amended the comment.